train/model.py

```diff
@@ -0,0 +1,218 @@
+
+        logits = self.pred_head(y)
+        return logits
+
+
+class Quantizer(nn.Module):
+    def __init__(self, n_embeddings, embedding_dim, commitment_cost):
+        super().__init__()
+        self.embedding_dim = embedding_dim
+        self.n_embeddings = n_embeddings
+        self.commitment_cost = commitment_cost
+
+        self.embedding = nn.Embedding(n_embeddings, embedding_dim)
+        self.embedding.weight.data.uniform_(-1/n_embeddings, 1/self.n_embeddings)
+
+    def compute_latent_loss(self, f_emb, quantized):
+        # Loss
+        e_latent_loss = F.mse_loss(quantized.detach(), f_emb)
+        q_latent_loss = F.mse_loss(quantized, f_emb.detach())
+        loss = q_latent_loss + self.commitment_cost * e_latent_loss
+        return loss
+
+    def forward(self, f_emb):
+        flat_input = f_emb.reshape(-1, self.embedding_dim)
+        # Calculate distances
+        distances = (torch.sum(flat_input**2, dim=1, keepdim=True) 
+                    + torch.sum(self.embedding.weight**2, dim=1)
+                    - 2 * torch.matmul(flat_input, self.embedding.weight.t()))
+            
+        # Encoding
+        encoding_indices = torch.argmin(distances, dim=1).unsqueeze(1)
+        encodings = torch.zeros(encoding_indices.shape[0], self.n_embeddings, device=f_emb.device)
+        encodings.scatter_(1, encoding_indices, 1)
+        
+        # Quantize and unflatten
+        quantized = torch.matmul(encodings, self.embedding.weight).reshape(f_emb.shape)
+        
+        quantized = f_emb + (quantized - f_emb).detach()
+        avg_probs = torch.mean(encodings, dim=0)
+        perplexity = torch.exp(-torch.sum(avg_probs * torch.log(avg_probs + 1e-10)))
+        
+        return quantized, perplexity, encoding_indices
+
+
+class VQVideo(nn.Module):
+    def __init__(self, n_dynamics_toks, n_frames, spatial_embeddings):
+        super().__init__()
+        self.n_dynamics_tokens = n_dynamics_toks
+        self.n_frames = n_frames
+
+        self.encoder = Encoder(
+            width=256,
+            layers=8,
+            heads=8,
+            n_tokens=n_dynamics_toks,
+            n_input_tokens=n_frames*128 + n_frames,
+            spatial_embeddings=spatial_embeddings,
+        )
+        self.decoder = Decoder(
+            width=256,
+            layers=8,
+            heads=8,
+            n_tokens=n_dynamics_toks,
+            n_input_tokens=n_dynamics_toks + N_FRAME_TOKS + 2,
+            spatial_embeddings=spatial_embeddings,
+        )
+        '''
+        self.quantizer = Quantizer(
+            n_embeddings=128,
+            embedding_dim=256,
+            commitment_cost=0.25,
+        )
+        '''
+
+    def encode_diff(self, x):
+        return self.encoder(x)
+
+    def decode(self, x, f):
+        logits = self.decoder(x, f)
+        # TODO: why does this matter???
+        true_logits = logits[:, :N_FRAME_TOKS]  # for now only one frame
+        # true_logits = logits[:, -N_FRAME_TOKS:]
```

Comment:
doesn't matter as much now after fixes and my hypothesis (at least for decoder) is because the model is residual you want the contextualized embeddings of x0 to look like spatial embeddings (somewhat) so if you take the ones from the beginning you will get taht since the embeddings below are from the spatial embedding table and not your new (and constantly changing) diff codebook